File: python/detectionformats/eventtype.py
#!/usr/bin/env python

# stdlib imports
import json


class EventType:
    """EventType - a conversion class used to create, parse, and validate
    event type data as part of detection data.
    """

    # json keys
    TYPE_KEY = "Type"
    CERTAINTY_KEY = "Certainty"

    def __init__(self, newType=None, newCertainty=None):
        """Initialize the event type object. Constructs an empty object
           if all arguments are None

        Args:
            newType: an optional String containing the event type to use
            newCertainty: an optional String containing the certainty to use
        Returns:
            Nothing
        Raises:
            Nothing
        """
        if newType is not None:
            self.type = newType
        if newCertainty is not None:
            self.certainty = newCertainty

    def fromJSONString(self, jsonString):
        """Populates the object from a json formatted string

        Args:
            jsonString: a required String containing the json formatted text
        Returns:
            Nothing
        Raises:
            Nothing
        """
        jsonObject = json.loads(jsonString)
        self.fromDict(jsonObject)

    def fromDict(self, aDict):
        """Populates the object from a dictionary

        Args:
            aDict: a required dictionary
        Returns:
            Nothing
        Raises:
            Nothing
        """
        try:
            self.type = aDict[self.TYPE_KEY]
            self.certainty = aDict[self.CERTAINTY_KEY]
        except (ValueError, KeyError, TypeError) as e:
            print("Dict format error, missing required keys: %s" % e)

    def toJSONString(self):
        """Converts the object to a json formatted string

        Args:
            None
        Returns:
            The JSON formatted message as a String
        Raises:
            Nothing
        """
        jsonObject = self.toDict()

        return json.dumps(jsonObject, ensure_ascii=False)

    def toDict(self):
        """Converts the object to a dictionary

        Args:
            None
        Returns:
            The dictionary
        Raises:
            Nothing
        """
        aDict = {}
        try:
            aDict[self.TYPE_KEY] = self.type
            aDict[self.CERTAINTY_KEY] = self.certainty
        except (NameError, AttributeError) as e:
            print("Missing required data error: %s" % e)

        return aDict

    def isValid(self):
        """Checks to see if the object is valid

        Args:
            None
        Returns:
            True if the object is valid, False otherwise
        Raises:
            Nothing
        """
        errorList = self.getErrors()

        return not errorList

    def getErrors(self):
        """Gets a list of object validation errors

        Args:
            None
        Returns:
            A List of Strings containing the validation error messages
        Raises:
            Nothing
        """
        errorList = []

        # optional values
        if hasattr(self, "type"):
            if self.type not in [
                "Earthquake",
                "MineCollapse",
                "NuclearExplosion",
                "QuarryBlast",
                "InducedOrTriggered",
                "RockBurst",
                "FluidInjection",
                "IceQuake",
                "VolcanicEruption",
            ]:
                errorList.append("Invalid Type in EventType Class.")

        if hasattr(self, "certainty"):
            if self.certainty not in ["Suspected", "Confirmed"]:
                errorList.append("Invalid Certainty in EventType Class.")

        return errorList

    def isEmpty(self):
        """Checks to see if the object is empty

        Args:
            None
        Returns:
            True if the object is empty, False otherwise
        Raises:
            Nothing
        """
        if hasattr(self, "type"):
            return False

        if hasattr(self, "certainty"):
            return False

        return True

File: java/src/main/java/gov/usgs/detectionformats/EventType.java
package gov.usgs.detectionformats;

import java.util.ArrayList;
import org.json.simple.JSONObject;

/**
 * a conversion class used to create, parse, and validate source data as part of detection data.
 * Detection formats supports a subset of the QuakeML 1.2 event types that are automatically
 * classifiable.
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class EventType implements DetectionInt {

  /** JSON Keys */
  public static final String TYPE_KEY = "Type";

  public static final String CERTAINTY_KEY = "Certainty";

  /** Optional event type */
  private final String type;

  /** Optional event type certantity */
  private final String certainty;

  /** The constructor for the EventType class. Initializes members to null values. */
  public EventType() {

    type = null;
    certainty = null;
  }

  /**
   * The advanced constructor for the EventType class. Initializes members to provided values.
   *
   * @param newType - A String containing the event type to use
   * @param newCertainty - A String containing the certainty to use
   */
  public EventType(String newType, String newCertainty) {

    type = newType;
    certainty = newCertainty;
  }

  /**
   * Constructs the class from a JSONObject, populating members
   *
   * @param newJSONObject - A JSONObject.
   */
  public EventType(JSONObject newJSONObject) {

    // optional values
    // type
    if (newJSONObject.containsKey(TYPE_KEY)) {
      type = newJSONObject.get(TYPE_KEY).toString();
    } else {
      type = null;
    }

    // certainty
    if (newJSONObject.containsKey(CERTAINTY_KEY)) {
      certainty = newJSONObject.get(CERTAINTY_KEY).toString();
    } else {
      certainty = null;
    }
  }

  /**
   * Converts the contents of the class to a JSONObject
   *
   * @return Returns a JSONObject containing the class contents
   */
  @SuppressWarnings("unchecked")
  public JSONObject toJSON() {

    JSONObject newJSONObject = new JSONObject();

    String jsonType = getType();
    String jsonCertainty = getCertainty();

    // required values
    // agencyID
    if ((jsonType != null) && (!jsonType.isEmpty())) {
      newJSONObject.put(TYPE_KEY, jsonType);
    }

    // certainty
    if ((jsonCertainty != null) && (!jsonCertainty.isEmpty())) {
      newJSONObject.put(CERTAINTY_KEY, jsonCertainty);
    }

    return (newJSONObject);
  }

  /**
   * Validates the class.
   *
   * @return Returns true if successful
   */
  public boolean isValid() {
    if (getErrors().size() == 0) {
      return (true);
    } else {
      return (false);
    }
  }

  /**
   * Gets any validation errors in the class.
   *
   * @return Returns a List&lt;String&gt; of any errors found
   */
  public ArrayList<String> getErrors() {
    String jsonType = getType();
    String jsonCertainty = getCertainty();

    ArrayList<String> errorList = new ArrayList<String>();

    // check for optional keys
    // type
    if ((jsonType != null) && (!jsonType.isEmpty())) {
      boolean match = false;
      if (jsonType.equals("Earthquake")) {
        match = true;
      } else if (jsonType.equals("MineCollapse")) {
        match = true;
      } else if (jsonType.equals("NuclearExplosion")) {
        match = true;
      } else if (jsonType.equals("QuarryBlast")) {
        match = true;
      } else if (jsonType.equals("InducedOrTriggered")) {
        match = true;
      } else if (jsonType.equals("RockBurst")) {
        match = true;
      } else if (jsonType.equals("FluidInjection")) {
        match = true;
      } else if (jsonType.equals("IceQuake")) {
        match = true;
      } else if (jsonType.equals("VolcanicEruption")) {
        match = true;
      } else {
        match = false;
      }

      if (!match) {
        // invalid eventType
        errorList.add("Invalid Type in EventType Class.");
      }
    }

    // certainty
    if ((jsonCertainty != null) && (!jsonCertainty.isEmpty())) {
      boolean match = false;
      if (jsonCertainty.equals("Suspected")) {
        match = true;
      } else if (jsonCertainty.equals("Confirmed")) {
        match = true;
      } else {
        match = false;
      }

      if (!match) {
        // invalid Certainty
        errorList.add("Invalid Certainty in EventType Class.");
      }
    }

    // success
    return (errorList);
  }

  /**
   * Checks to see if this object is empty
   *
   * @return Returns true if empty, false otherwise.
   */
  public boolean isEmpty() {
    if (getType() != null) {
      return (false);
    }
    if (getCertainty() != null) {
      return (false);
    }
    return (true);
  }

  /** @return the type */
  public String getType() {
    return type;
  }

  /** @return the certainty */
  public String getCertainty() {
    return certainty;
  }
}
